### ISSUE TITLE
training question

### ISSUE THREAD

**Issue by youjin-c** [snap-research/CAT#24]
Hello again,

I am training CycleGAN with ffhq dataset, trying to make a filter that makes the user smile.
I checked with my training recently, but the result wasn't satisfying.
So, I'd like to ask for advice from you to enhance the performance of the model.

Initially, I trained with 2500 unaligned pairs of smiling/not smiling 1024x1024 photos and got `fid_B: 88.244` with `--nepochs 500 --nepochs_decay 500`
While training the teacher model, I found the best model is found on epoch 181 / 1000. I kinda kept training till the 841st epoch but couldn't find the better one.
For distilling, I got `fid: 87.976` at epoch 270 with these options `--nepochs 500 --nepochs_decay 500`

![Screen Shot 2022-09-13 at 4 33 05 PM](https://user-images.githubusercontent.com/18349962/190733957-2ed48d9b-197d-4370-b498-88489d6788f6.png)
![Screen Shot 2022-09-13 at 4 33 14 PM](https://user-images.githubusercontent.com/18349962/190734125-a9be5e5f-db70-48aa-9a00-735e097ea892.png)

So I added 2500 more unaligned pairs to the dataset and am training again now.
I am eager to make this filter look great. Any advisory is appreciated.

Also, one more question is, is it right to resume with the latest_net, not the best_net I have?
With the same learning rate options, I guess resuming from the best net, or 180_net again, is just repeating what I've trained from the last training.
Please let me know if my assumption is correct.

**Comment by alanspike** [comment]
Hi @youjinChung, we found the FID might not be the best metric to evaluate the image quality for all the datasets/tasks. If you prefer to use FID, then you might consider [clean-fid](https://github.com/GaParmar/clean-fid). 

Or you could simply check the results (generated images) saved during the training time to see which one gives you the best visual results. Then you could use that specific checkpoint instead of the `best` or the `latest`.

Before doing the model compression, it is helpful to check whether the teacher model gives you good results. It is more reasonable to do the model compression/distillation when the teacher model is good enough. 

Another thing is that you train the model on the FFHQ dataset, which has the face images well aligned. So to get more reasonable testing results, you might also want to have the testing facial images aligned and apply the model to face only.

**Comment by youjin-c** [comment]
Hello @alanspike,

thank you for such a quick response.
So I can try 
1. Implement clean-fid instead of fid, or 
2. Check eval images to select the model or 
3.  Since FFHQ is too well aligned  with the facial landmark positons, I could try to make a dataset with a randomized alignment? Such as rotate, flip and resize the faces randomly?
Could you confirm if my understanding is correct?

**Comment by alanspike** [comment]
You don't have to try clean-fid if you feel comfortable to manually check some checkpoints to select the best one.

I don't think you need to do much about the training dataset. However, for the inference time, instead of applying the model on the whole image that includes human body, you could try only apply the model on the human face.

**Comment by youjin-c** [comment]
Gotcha.
Sadly those images are already applied to the face only. I fed face crop texture as input.

**Comment by youjin-c** [comment]
Hi,

So I selected a teacher model by checking eval files visually. They are not the best but look good.
![00008](https://user-images.githubusercontent.com/18349962/191292998-cad758f8-33d1-497c-9271-7e52650cae16.png)
![00013](https://user-images.githubusercontent.com/18349962/191292982-e0cbc83a-1943-4aa8-bf5f-58bb4c3a05f5.png)
![00015](https://user-images.githubusercontent.com/18349962/191292971-6e00d557-bee0-4e62-b892-1c21d5c8adf4.png)

so the images were under folder `183902 (CAT/logs/cycle_gan/face2smile/inception/teacher/eval/183902/AtoB/fake/00015.png),`and I found that epoch 650 is iters 183902 from the log.
```
End of epoch 650 / 1005 	 Time Taken: 638.88 sec
###(Evaluate epoch: 650, iters: 183902, time: 94.290) fid_B: 75.502 fid_B-mean: 76.993 fid_B-best: 72.889 fid_A: 78.560 fid_A-mean: 77.804 fid_A-best: 75.231 
Saving the model at the end of epoch 650, iters 183902
learning rate = 0.0001686
```

So I started distilling with 650 model as teacher, but the Tfake look so different from the eval images of training.
![00015 (1)](https://user-images.githubusercontent.com/18349962/191294368-0e77a98a-51f3-4208-a41f-c1daef331003.png)
![00013 (1)](https://user-images.githubusercontent.com/18349962/191294378-2a3e50c9-bfe3-4380-8eda-382f49735347.png)
![00008 (1)](https://user-images.githubusercontent.com/18349962/191294400-d328f26e-b85e-4846-8a1b-92292cd5906e.png)

**_I wonder what makes these differences,_** and I assume this affects the distilling into the student model. 

My assumption is my dataset is 1024x1024, and I preprocessed the dataset for distilling, which affected the inference of the teacher model.
```
  --preprocess resize_and_crop \
  --load_size 512 \
  --crop_size 512,512 \
```
and I didn't find any preprocessing options for `get_real_stat.py` or `train.py`, but now the training eval images look like 256x256. so maybe scaling to 512x512 from 256x256 made this change?

Since the teacher doesn't infer as expected, the student model is also not being trained well. Please let me know how to fix this issue.
I am thinking of resizing and cropping to 256x256 but leave this question here first.
One concern is that I wanted my model to be high resolution, but what I did is already on 256x256, and It won't look high res on Snap. Is it right?

Below are the whole commands for training and distilling.
```
!python train.py --dataroot database/face2smile \
  --model cycle_gan \
  --log_dir logs/cycle_gan/face2smile/inception/teacher \
  --netG inception_9blocks \
  --real_stat_A_path real_stat/face2smile_A.npz \
  --real_stat_B_path real_stat/face2smile_B.npz \
  --batch_size 16 \
  --num_threads 32 \
  --gpu_ids 0,1,2,3 \
  --norm_affine \
  --norm_affine_D \
  --channels_reduction_factor 6 \
  --kernel_sizes 1 3 5 \
  --restore_G_A_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/latest_net_G_A.pth\
  --restore_D_A_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/latest_net_D_A.pth\
  --restore_G_B_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/latest_net_G_B.pth\
  --restore_D_B_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/latest_net_D_B.pth\
  --save_latest_freq 10000 --save_epoch_freq 5 \
  --epoch_base 586 --iter_base 163362 \
  --nepochs 0 --nepochs_decay 420
```

```
!python distill.py --dataroot database/face2smile \
  --dataset_mode unaligned \
  --distiller inception \
  --gan_mode lsgan \
  --log_dir logs/cycle_gan/face2smile/inception/student \
  --restore_teacher_G_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/650 _net_G_A.pth \
  --restore_pretrained_G_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/650 _net_G_A.pth \
  --restore_D_path logs/cycle_gan/face2smile/inception/teacher/checkpoints/650 _net_D_A.pth \
  --real_stat_path real_stat/face2smile_B.npz \
  --teacher_netG inception_9blocks --student_netG inception_9blocks \
  --pretrained_ngf 64 --teacher_ngf 64 --student_ngf 20 \
  --ndf 64 \
  --num_threads 80 \
  --eval_batch_size 4 \
  --batch_size 40 \
  --gpu_ids 0,1,2,3 \
  --norm_affine \
  --norm_affine_D \
  --channels_reduction_factor 6 \
  --kernel_sizes 1 3 5 \
  --lambda_distill 1.0 \
  --lambda_recon 5 \
  --prune_cin_lb 16 \
  --target_flops 2.6e9 \
  --distill_G_loss_type ka \
  --save_epoch_freq 1 \
  --save_latest_freq 500 \
  --norm_student batch \
  --padding_type_student zero \
  --norm_affine_student \
  --norm_track_running_stats_student \
  --preprocess resize_and_crop \
  --load_size 512 \
  --crop_size 512,512 \
  --nepochs 500 --nepochs_decay 500
```

**Comment by youjin-c** [comment]
So yes,
when I disillted with 256x256, Tfake and Sfake looked normal.
![image](https://user-images.githubusercontent.com/18349962/191343594-26958c1d-d5ba-43d5-8255-6c6df9a2b881.png)
![image](https://user-images.githubusercontent.com/18349962/191343631-45e3b1ce-5823-4849-a21b-db1dbd1d9a72.png)
(For future users who might interested)
so `--preprocess` option is BOTH on `base_options.py` and `get_real_stat.py`, which is applied to train as default as well. (OMG I am too late to notice this)
